docs/asdf/developer_overview.rst

```diff
@@ -350,3 +354,209 @@ which in turn calls ``_make_schema_loader``, then ``_load_schema``.
+to extension developers, but it was discarded because it wasn't thougth
+it was worth turning all those high performance containers into slower
+asdf subclasses.  We may want to revisit this if we decide to implement
+a tree that tracks "dirty" nodes and only writes to disk those that
+have changed, since in that case we'll need custom container subclasses
+anyway.  We could also consider writing our own dict/list subclass in C
+so we could have our cake and eat it too.
+
+The ``walk_and_modify`` code handles the case where the tag code returns 
+a generator instead of a value. These generators are expected to be the
+similar kind of generator that pyyaml uses, but differing in that instead
+of returing an empty container object it will populate whatever elements
+it can completely (e.g, all non-mutable ones), and completing the 
+population of all the mutable memebers on the second iteration
+(which may, in turn, generate new generators for mutable elements 
+contained within). In handling these generators, the 
+first iteration also saves the generator in the context. When the
+top level of the context is reached (it handles nesting by indicating
+how many times it has been entered as a context), it starts "draining"
+the saved generators by doing the second iteration on them. Like 
+pyyaml, this second iteration may produce yet more generators that
+get saved, and thus keeps iterating on the saved generators until none
+are left.
+
+It is not possible to construct reference cycles in immutable
+objects within pure Python code, and thus the generators are only needed
+for mutable constructs (e.g., dicts and lists).
+
+Historical note: versions of the ASDF library prior to 2.6.0 required
+tag code when converting from a tagged object to a custom object to 
+call ``tagged_tree_to_custom_tree`` on any values of attributes that may be
+arbitrarily nested objects. That no longer is needed with the latest code
+since any attribute that contains a mapping or sequence object automatically
+uses a generator, so population of that attribute is automatically
+deferred until the context is exited. Thus there is no need to explicitly
+call a function to populate it.
+
+More explicitly, the ``_recurse`` function defined within the tree walker
+(in the usual postorder case) calls ``_handle_children()`` of the node
+in question first. If the node consists of a mutable container, that
+itself calls _recurse and collected with the results of all the children
+and the resulting structure is yielded. The effect is that a tree is
+returned consisting of scalars, tuples, and empty mutable items. This
+result is passed along to the callback, i.e., the ``tag.from_tree_tagged()``
+method. Since any of its attributes that itself is mutable is an empty
+container awaiting its registered generator to be called, there is 
+no need to explicitly populate it since when the generators are drained,
+it will be populated.
+
+Thus reader, your mind shall now be drained.
```

Comment:
Oh I bet we have a few iterations left in us...

Reply:
How many is indeterminate though.